CAM: 3D Surface Enhance Performance

Dimitris75 opened the pull request:
### **PR Summary:  Enhance Performance and Simplify 3D Surface UI**

 This pull request introduces a significant user experience (UX) and performance enhancement for the 3D Surface operation. It simplifies the complex mesh conversion settings into a single, intuitive "Speed vs Accuracy" slider. Crucially, it also introduces Mesh Simplification as a new core feature of the operation, providing a powerful new way to accelerate path calculation on complex models.

### **Benefits and Rationale**
The primary goal of these changes is to make the 3D Surface operation more accessible and efficient for all users.

**_Simplified User Experience:_** 
 The previous interface required users to understand two separate technical parameters (AngularDeflection, LinearDeflection). The new UI replaces this with a single, goal-oriented slider. Users can now make an intuitive choice between faster processing or higher accuracy without needing expert knowledge.

**_Faster Path Calculation via New Mesh Simplification Feature:_**

- _New Capability:_ 

A new MeshSimplification property has been added to the operation. This powerful feature, now controlled by the slider, utilizes a decimation algorithm to intelligently reduce the triangle count of the model's mesh before it is sent to the main OCL algorithm. This is a new optimization that dramatically reduces the data load and is especially effective on complex, organic shapes.

- _Performance Impact:_ 

 On "Speed" settings, this new feature can reduce the triangle count by a significant margin, leading to an order-of-magnitude reduction in computation time. This enables a rapid, iterative workflow that was not previously possible on complex models.

_**More Efficient Machining Time:**_
 The generated G-code is now more efficient for multi-pass operations. By implementing "air cutting" avoidance, the tool no longer makes slow feed-rate moves over areas that have already been cleared, significantly reducing on-machine cycle time.


### **Technical Implementation Details**
The work is divided between UI modifications and backend logic enhancements, including the integration of the new Mesh Simplification feature.

1. **UI and Property Management** (PageOpSurfaceEdit.ui, Gui/Surface.py, Op/Surface.py)

- _UI Layout (.ui):_ The Task Panel has been simplified to feature a single QSlider (accuracySlider) within a "Speed vs Accuracy" group, replacing the previous manual input fields. The slider is configured to be responsive to panel resizing.

- _Backend Properties (Op/Surface.py):_

- [ ] A new App::PropertyFloat named MeshSimplification was added to the operation.

- [ ] A new App::PropertyInteger named Accuracy was introduced to store the slider's 1-7 value.

- [ ] The AngularDeflection, LinearDeflection, and the new MeshSimplification properties are now hidden from the Properties View (setEditorMode(2)) to ensure they are only controlled via the slider.

- [ ] The _applyAccuracySettings method serves as the core engine for this feature. It contains a dictionary that translates the slider's Accuracy value into the concrete numerical values for all three hidden mesh properties.

- _UI Controller (Gui/Surface.py):_ This file was simplified to connect the accuracySlider to the backend Accuracy property.

2. **Performance Optimizations** (Backend)

- _Integration of Mesh Simplification_ (_makeSTL in SurfaceSupport.py): This function was significantly enhanced to implement the new Mesh Simplification feature.

- [ ] _Decimation (New Feature):_ The function now applies the mesh.decimate() algorithm based on the MeshSimplification value derived from the slider. This is the core of the new capability, providing a powerful tool to reduce mesh density and accelerate calculation.

- [ ] _Pre-Clipping:_ The function also performs a pre-clipping boolean operation by using the Final Depth specified to ensure only the relevant portion of the model is passed to the computationally expensive tessellation and decimation steps. * The model is clipped below the final depth specified.

- _Intelligent Multi-Pass_ (_planarDropCutMulti in Surface.py): The logic for multi-pass operations was fundamentally improved. It now analyzes the path for each layer against the depth of the previously cleared layer and prunes "air cutting" segments, replacing them with efficient G0 rapid moves.

- _Lightweight Collision Model_ (_makeSafeSTL in SurfaceSupport.py): To speed up the calculation of safe rapid moves, this function now creates a hollow, low-polygon model by extracting only the outer shells (fused.Shells) of the safety geometry, which is significantly faster for the collision checker to process.

### **Conclusion**
In summary, these changes deliver a more intuitive, performant, and efficient 3D Surface operation. The key outcomes of this project are:

- **Vastly Improved Usability:** The "Speed vs Accuracy" slider simplifies the user experience.

- **Faster Path Generation:** The introduction of the Mesh Simplification feature, combined with other optimizations, provides tangible reductions in user wait time.

- **More Efficient Machining:** By eliminating redundant "air cutting" moves, the generated G-code is more efficient, leading to shorter cycle times on the CNC machine.

<img width="520" height="346" alt="Screenshot From 2025-10-05 21-02-14" src="https://github.com/user-attachments/assets/5b3778be-7e19-4092-8ec5-3a42bfe2b264" />

<img width="1414" height="685" alt="Screenshot From 2025-10-05 21-07-13" src="https://github.com/user-attachments/assets/d951fb29-e220-41da-a75a-5697e477f17a" />

Extensive testing has been conducted on a diverse range of models, from simple prismatic parts with flat faces to complex, organic shapes. The primary focus of this testing was to validate the reliability of the new algorithms and to ensure the user experience of the new "Speed vs Accuracy" slider is both intuitive and effective.

 While the results have been consistently positive—demonstrating significant performance gains and robust path generation—wider feedback from the community is encouraged and essential for identifying any remaining edge cases and ensuring long-term stability.

Dimitris75 committed:
3D Surface Enhance Performance

Simplify 3D Surface UI and Enhance Performance

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

hyarion commented:
Looks like both code and description is AI generated, which makes me wonder:

* Is it backwards compatible?
* Does it work?
* Is it as tested as the LLM claims it is?
* Does it make sense?

Angular and linear deflection did mean something to the user (though not really self explanatory). Accuracy vs. speed doesn't really explain what is happening, does it?

Dimitris75 committed:
Correct messages and comments

Correct messages and comments

Dimitris75 committed:
Disable clipping for Rotational

Disable clipping for Rotational Scan Type

Dimitris75 committed:
Remove unused local variable

Remove unused local variable

Dimitris75 committed:
Fix compatibility with Waterline issue

Maintain compatibility with Waterline operation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Dimitris75 committed:
Remove temporary parameters from Mesh

Remove forgotten testing parameters on makeSTL

Dimitris75 committed:
Merge branch '3D-Surface-Enhance-Performance' of https://github.com/Dimitris75/FreeCAD into 3D-Surface-Enhance-Performance

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

maxwxyz commented:
The PR has a conflict.
@sliptonic @Connor9220 should I add it to the CAM board?

Dimitris75 committed:
Update to the latest Main

Update to the latest main changes.

Dimitris75 committed:
Update

Dimitris75 committed:
Merge branch 'main' into 3D-Surface-Enhance-Performance

Dimitris75 committed:
Fix UI panel

Fix UI panel

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Dimitris75 committed:
Update PageOpSurfaceEdit.ui

Update PageOpSurfaceEdit.ui file

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Dimitris75 committed:
Remove unnecessary pass and comments

Remove unnecessary pass and comments

Dimitris75 committed:
Merge branch '3D-Surface-Enhance-Performance' of https://github.com/Dimitris75/FreeCAD into 3D-Surface-Enhance-Performance

Dimitris75 committed:
Set new Default Accuracy Values and Limits

-Set new Default Accuracy Values
-Set Accuracy limits
-Update Accuracy Settings and fix console message

Dimitris75 committed:
Fix Optimize stepover transition

Remove remainings from early attempts to fix multi-pass and fix Optimize stepover transitions

Dimitris75 committed:
Remove cutter dia restriction

Remove cutter dia restriction from StepOver Transition Optimization for short distances.  The safePDC has already calculated that this path is safe.

Dimitris75 committed:
Correct small typo error

Correct small typo error

Dimitris75 committed:
Fix Multi-pass Corner Bug in

Replace the logic that was "auto-shifting" the Z-height with logic that explicitly uses DepthOffset.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Dimitris75 committed:
Remove forgotten unnecessary assignment to 'zs'.

Remove forgotten unnecessary assignment to 'zs'.

Dimitris75 committed:
Remove helper function getTransition

The standard layer processing loop (_planarDropCutSingle) already calls _stepTransitionCmds, which automatically and safely handles transitions

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Dimitris75 commented:
@sliptonic 

I have a question. I haven't seen that before, so I suspect that the answer is no. However, can I import os in the SurfaceSupport.py file? 

In the following code below, we are using a slow Python loop to transfer our Mesh to OCL, triangle by triangle. 
https://github.com/FreeCAD/FreeCAD/blob/776b2f2d5ac64e0d5150ede18809e22e3a9d6125/src/Mod/CAM/Path/Op/SurfaceSupport.py#L1290-L1299

If I use mesh.write(C++) to a temp file instead, then I can instantly load that file from OCL.

Dimitris75 commented:
Cool! Yes, I will use a fallback just in case.

Dimitris75 committed:
Redefine accuracy map

Mesh Simplification to a maximum of 15%.

Dimitris75 committed:
Remove multiple console messages

Remove multiple console messages for layer cutting depth

Dimitris75 committed:
Include  changes from Fix Rotational Scan

Include  changes from Fix Rotational Scan

Dimitris75 committed:
Add Support for taperedballnose

Add Support for taperedballnose

Dimitris75 commented:
@Connor9220 

Yes, that is coming from your PR (Already merged). I'm trying to keep this PR up to date.

Dimitris75 committed:
Merge branch 'main' into 3D-Surface-Enhance-Performance

Dimitris75 committed:
Remove garbage from conflict resolving

Remove garbage from conflict resolving

hyarion commented on src/Mod/CAM/Path/Op/Surface.py:
```diff
@@ -1415,6 +1484,21 @@ def _planarDropCutSingle(self, JOB, obj, pdc, safePDC, depthparams, SCANDATA):
+                unique_points.append(points[i])
+
+        # All subsequent operations now use the clean, de-duplicated list.
+        points = unique_points
```
If you import groupby (`from itertools import groupby`)

Then this can be rewritten as:
```suggestion
    points = [p for p, _ in groupby(points)]
```

hyarion commented on src/Mod/CAM/Path/Op/Surface.py:
```diff
@@ -2537,15 +2407,22 @@ def _getTransitionLine(self, pdc, p1, p2, obj):
+            minZ = maxZ = 0.0
+        else:
+            minZ = min(zs)
+            maxZ = max(zs)
```
```suggestion
        zs = [p.z for p in pdcLine]
        minZ = min(zs, default=0.0)
        maxZ = max(zs, default=0.0)
```

hyarion commented on src/Mod/CAM/Path/Op/SurfaceSupport.py:
```diff
@@ -1277,13 +1228,71 @@ def _makeSTL(model, obj, ocl, model_type=None):
+        if hasattr(obj, "MeshSimplification") and obj.MeshSimplification > 0:
+            reduction_percent = obj.MeshSimplification
+            if reduction_percent > 99.0:
+                reduction_percent = 99.0  # Safety cap
```
```suggestion
            allowed_max = 99.0
            reduction_percent = min(obj.MeshSimplification, allowed_max)
```

Dimitris75 commented on src/Mod/CAM/Path/Op/Surface.py:
```diff
@@ -758,6 +800,15 @@ def opExecute(self, obj):
+            f"Mesh Simplification = {obj.MeshSimplification:.1f}%\n"
+        )
+        FreeCAD.Console.PrintMessage(accuracy_msg)
+
```
Thank for your review. That's mostly for "power users" but, maybe we can squeeze a few extra info under the slider. All of your suggestions looking good to me!

Dimitris75 committed:
Small enhancements and simplifications

Small enhancements and simplifications

Dimitris75 commented on src/Mod/CAM/Path/Op/SurfaceSupport.py:
```diff
@@ -413,8 +413,7 @@ def _extractOffsetFaces(self):
             for f in offsetArea.Faces:
-                wires = wires + f.Wires
-                # wires.extend(f.Wires)
+                wires.extend(f.Wires)
```
Yes, I already did.

hyarion replied:
@Dimitris75 in the code you're still iterating the faces with a for loop, which calls extends once per face.
With @tarman3 suggestion you would generate an intermediate list of all wires in all faces in `offsetArea`. Or did you do that but haven't pushed it yet?

Dimitris75 replied:
Do be honest, I prefer the for loop. Although tarman's version is more Pythonic, the for loop is easier to read and provides room for additional features if needed. The tiny fraction of a millisecond saved by a list comprehension in the extraction step is irrelevant to the overall recompute time.

tarman3 committed:
minor fixes

tarman3 commented:
@Dimitris75 
Please take a look
https://github.com/Dimitris75/FreeCAD/pull/1

Dimitris75 committed:
Merge pull request #1 from tarman3/mathceil_fix

minor fixes

hyarion commented:
What is the status on this PR?
Are we waiting on DWG, CAM people, or the author?

sliptonic commented:
> What is the status on this PR? Are we waiting on DWG, CAM people, or the author?

No.  The PR won't be merged.  The work has shifted to a reimplementation of surface into two separate operations.
A planar surfacing op with multiple strategies and a rotary surfacing op also with multiple strategies.
I'm leaving the PR in draft because it contains some improvements that I want to make sure get incorporated into the new ops.
